gazebo_ros/src/qos.cpp

```diff
@@ -0,0 +1,219 @@
+    } else if ("manual_by_topic" == liveliness) {
+      qos_overrides.liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC;
+    } else {
+      std::stringstream oss;
+      oss << "invalid setting for liveliness '" << liveliness << "'";
+      throw InvalidQoSException(oss.str());
+    }
+  }
+
+  // Parse 'liveliness_lease_duration' QoS
+  if (_sdf->HasElement("liveliness_lease_duration")) {
+    auto lease = _sdf->GetElement("liveliness_lease_duration")->Get<size_t>();
+    qos_overrides.liveliness_lease = std::chrono::milliseconds(lease);
+  }
+
+  return qos_overrides;
+}
+
+QoS::QoS(sdf::ElementPtr _sdf)
+{
+  // If there's no <qos> tag, then there's nothing to do
+  if (!_sdf->HasElement("qos")) {
+    return;
+  }
+  sdf::ElementPtr qos_sdf = _sdf->GetElement("qos");
+
+  // If there's no <topic> tags, then there's nothing to do
+  if (!qos_sdf->HasElement("topic")) {
+    return;
+  }
+  sdf::ElementPtr topic_sdf = qos_sdf->GetElement("topic");
+
+  while (topic_sdf) {
+    if (!topic_sdf->HasAttribute("name")) {
+      throw InvalidQoSException("topic element missing required 'name' attribute");
+    }
+    auto topic_name = topic_sdf->Get<std::string>("name");
+
+    // For each topic, get the publisher QoS overrides
+    if (topic_sdf->HasElement("publisher")) {
+      publisher_qos_overrides_map_.emplace(
+        topic_name, QoS::get_qos_overrides_from_sdf(topic_sdf->GetElement("publisher")));
+    }
+
+    // For each topic, get the subscription QoS overrides
+    if (topic_sdf->HasElement("subscription")) {
+      subscription_qos_overrides_map_.emplace(
+        topic_name, QoS::get_qos_overrides_from_sdf(topic_sdf->GetElement("subscription")));
+    }
+
+    topic_sdf = topic_sdf->GetNextElement("topic");
+  }
+}
+
+rclcpp::QoS QoS::apply_overrides(const QoS::QoSOverrides & overrides, const rclcpp::QoS default_qos)
+{
+  rclcpp::QoS qos = default_qos;
+  if (overrides.reliability != RMW_QOS_POLICY_RELIABILITY_UNKNOWN) {
+    qos.reliability(overrides.reliability);
+  }
+
+  if (overrides.durability != RMW_QOS_POLICY_DURABILITY_UNKNOWN) {
+    qos.durability(overrides.durability);
+  }
+
+  if (overrides.history == RMW_QOS_POLICY_HISTORY_KEEP_ALL) {
+    qos.keep_all();
+  } else if (overrides.history == RMW_QOS_POLICY_HISTORY_KEEP_LAST) {
+    qos.keep_last(overrides.depth);
+  }
+
+  if (overrides.liveliness != RMW_QOS_POLICY_LIVELINESS_UNKNOWN) {
+    qos.liveliness(overrides.liveliness);
+  }
+
+  if (overrides.deadline != std::chrono::milliseconds::zero()) {
+    qos.deadline(overrides.deadline);
+  }
+
+  if (overrides.lifespan != std::chrono::milliseconds::zero()) {
+    qos.lifespan(overrides.lifespan);
+  }
+
+  if (overrides.liveliness_lease != std::chrono::milliseconds::zero()) {
+    qos.liveliness_lease_duration(overrides.liveliness_lease);
+  }
+
+  return qos;
+}
+
+
+rclcpp::QoS QoS::get_publisher_qos(const std::string topic, rclcpp::QoS default_qos) const
```

Comment:
We will need to make use of this function and `get_subscription_qos` in all plugins, right?

Reply:
That's right. See the connected PR for usage: https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1092